proof-systems/src/darlin/poly-commit/src/ipa_pc/constraints/mod.rs

```diff
@@ -0,0 +1,350 @@
+use crate::ipa_pc::constraints::data_structures::{
+    BulletPolynomial, IPAMultiPointProof, IPAProofGadget, IPAVerifierKeyGadget, IPAVerifierState,
+};
+use crate::ipa_pc::InnerProductArgPC;
+use crate::{Error, PolynomialCommitmentVerifierGadget};
+use algebra::{EndoMulCurve, Field, PrimeField};
+use fiat_shamir::constraints::FiatShamirRngGadget;
+use fiat_shamir::FiatShamirRng;
+use r1cs_core::{ConstraintSystemAbstract, SynthesisError};
+use r1cs_std::boolean::Boolean;
+use r1cs_std::fields::fp::FpGadget;
+use r1cs_std::fields::nonnative::nonnative_field_gadget::NonNativeFieldGadget;
+use r1cs_std::fields::FieldGadget;
+use r1cs_std::groups::EndoMulCurveGadget;
+use r1cs_std::to_field_gadget_vec::ToConstraintFieldGadget;
+use r1cs_std::FromBitsGadget;
+use std::marker::PhantomData;
+use rand_core::SeedableRng;
+use rand_xorshift::XorShiftRng;
+
+mod data_structures;
+
+#[cfg(feature = "circuit-friendly")]
+#[cfg(test)]
+mod tests;
+
+/// Helper function that allows to compute `base_point`*`scalar` by dealing also with the case that
+/// `base_point` is zero. The `endo_mul` flag specifies if the multiplication between `base_point`
+/// and `scalar` must be performed with `endo_mul` or a simple `mul_bits`
+pub(crate) fn safe_mul<'a, ConstraintF, G, GG, CS, IT>(
```

Comment:
Not sure if I got what you mean, the `endo_mul` flag is not equivalent to the `circuit-friendly` feature. The `safe_mul` is employed to multiply a base point to  a scalar with either `endo_mul` or `mul_bits`, depending on the case: `endo_mul` will be employed only when the `scalar` is a squeezed challenge, while in all the other cases we need to employ `mul_bits`. Therefore, `mul_by_challenge` is not a wrapper to `safe_mul`. However, since I have also introduced a `safe_mul_bits` function for the PCGadget, I realized that the flag for `safe_mul` is no longer needed: instead of calling `safe_mul` with `endo_mul=false`, we can directly call `safe_mul_bits`. Therefore, I can remove the `safe_mul` function and move its code to the `mul_by_challenge`, which always perform an `endo_mul`.